File: 1450.number-of-students-doing-homework-at-a-given-time.cpp
/*
 * @lc app=leetcode id=1450 lang=cpp
 *
 * [1450] Number of Students Doing Homework at a Given Time
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  int busyStudent(vector<int>& startTime, vector<int>& endTime, int queryTime) {
    int size = startTime.size();
    int ret = 0;
    for (int i = 0; i < size; i++) {
      if (startTime[i] <= queryTime && queryTime <= endTime[i]) {
        ret++;
      }
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> startTime, endTime;
  startTime = {1, 2, 3};
  endTime = {3, 2, 7};
  ASSERT_EQ(s.busyStudent(startTime, endTime, 4), 1);
  startTime = {4};
  endTime = {4};
  ASSERT_EQ(s.busyStudent(startTime, endTime, 4), 1);
  return 0;
}

File: 1450.number-of-students-doing-homework-at-a-given-time.rs
/*
 * @lc app=leetcode id=1450 lang=rust
 *
 * [1450] Number of Students Doing Homework at a Given Time
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn busy_student(start_time: Vec<i32>, end_time: Vec<i32>, query_time: i32) -> i32 {
        let mut ans = 0;
        for (i, s) in start_time.iter().enumerate() {
            if *s <= query_time && query_time <= end_time[i] {
                ans += 1;
            }
        }
        ans
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::busy_student(vec![1, 2, 3], vec![3, 2, 7], 4), 1);
}

#[test]
fn test2() {
    assert_eq!(Solution::busy_student(vec![4], vec![4], 4), 1);
}
